mono/metadata/handle.h

```diff
@@ -16,183 +18,231 @@
-	} while (0)
+		CLEAR_STACK_WATERMARK	\
+		CLEAR_ICALL_COMMON	\
+		void* __ret = MONO_HANDLE_RAW (HANDLE);	\
+		CLEAR_ICALL_FRAME	\
+		return __ret;	\
+	} while (0); } while (0)
 
-#define MONO_HANDLE_SETREF(handle,fieldname,value)			\
-	do {								\
-		MonoHandle __value = (MonoHandle) (value);		\
-		MONO_PREPARE_GC_CRITICAL_REGION;					\
-		MONO_OBJECT_SETREF (mono_handle_obj ((handle)), fieldname, mono_handle_obj (__value)); \
-		MONO_FINISH_GC_CRITICAL_REGION;					\
-	} while (0)
+//Handle macros/functions
 
-#define MONO_HANDLE_SETREF_NULL(handle,fieldname)			\
-	do {								\
-		MONO_PREPARE_GC_CRITICAL_REGION;			\
-		MONO_OBJECT_SETREF (mono_handle_obj ((handle)), fieldname, NULL); \
-		MONO_FINISH_GC_CRITICAL_REGION;				\
-	} while (0)
+//FIXME respect the CHECKED_BUILD define
+#define HANDLE_INVARIANTS(H) mono_handle_verify((void*)(H).__obj)
 
+#define TYPED_HANDLE_NAME(__T) __T ## Handle
+#define TYPED_HANDLE_DECL(__T, __SN) \
+	typedef struct { __T **__obj; } TYPED_HANDLE_NAME (__T) ;
 
-#define MONO_HANDLE_SET(handle,fieldname,value)	\
-	do {	\
-		MONO_PREPARE_GC_CRITICAL_REGION;	\
-		mono_handle_obj ((handle))->fieldname = (value);	\
-		MONO_FINISH_GC_CRITICAL_REGION;	\
-	} while (0)
+#define MONO_HANDLE_INIT { (void*)mono_null_value_handle.__obj }
+#define NULL_HANDLE mono_null_value_handle
 
-#define MONO_HANDLE_ARRAY_SETREF(handle,index,value)			\
-	do {								\
-		MonoHandle __value = (MonoHandle) (value);		\
-		MONO_PREPARE_GC_CRITICAL_REGION;					\
-		mono_array_setref (mono_handle_obj ((handle)), (index), mono_handle_obj (__value)); \
-		MONO_FINISH_GC_CRITICAL_REGION;					\
-	} while (0)
+//XXX add functions to get/set raw, set field, set field to null, set array, set array to null
+#define MONO_HANDLE_RAW(H) (HANDLE_INVARIANTS (H), (*(H).__obj))
+#define MONO_HANDLE_DCL(type, name) TYPED_HANDLE_NAME(type) name = { mono_handle_new ((MonoObject*)(name ## _raw)) }
+#define MONO_HANDLE_NEW(TYPE, VALUE) (TYPED_HANDLE_NAME(TYPE)){ mono_handle_new ((MonoObject*)(VALUE)) }
 
-#define MONO_HANDLE_ARRAY_SETREF_NULL(handle,index)			\
-	do {								\
-		MONO_PREPARE_GC_CRITICAL_REGION;			\
-		mono_array_setref (mono_handle_obj ((handle)), (index), NULL); \
-		MONO_FINISH_GC_CRITICAL_REGION;				\
+/*
+WARNING WARNING WARNING
+
+The following functions require a particular evaluation ordering to ensure correctness.
+We must not have exposed handles while any sort of evaluation is happening as that very evaluation might trigger
+a safepoint and break us.
+
+This is why we evaluate index and value before any call to MONO_HANDLE_RAW or other functions that deal with naked objects.
+*/
+#define MONO_HANDLE_SETRAW(HANDLE, FIELD, VALUE) do {	\
+		MonoObject *__val = (MonoObject*)(VALUE);	\
+		MONO_OBJECT_SETREF (MONO_HANDLE_RAW (HANDLE), FIELD, __val);	\
+	} while (0);
+
+#define MONO_HANDLE_SET(HANDLE, FIELD, VAL_HANDLE) do {	\
+		MONO_OBJECT_SETREF (MONO_HANDLE_RAW (HANDLE), FIELD, MONO_HANDLE_RAW (VAL_HANDLE));	\
```

Comment:
Isn't this exactly what your comment says we shouldn't have? we'll have the raw ptr from `HANDLE` on the stack while we evaluate `VAL_HADNLE`